<h2>협업 필터링 -collaborative Filtering</h2>

말 그대로 협업을 하는거임. 추천시스템에 웬 협업? 이렇게 생각할 수 있는데

쉽게 이해할 수 있다. 영화를 보러갈때 보통 어떻게 하나? 친구한테 물어본다.

이 영화 어때? 친구가 "좋아"라고 한다면 볼만하다. 유의할 점은 취향이 비슷한 친구한테 물어봐야된다는 것임.

<img src="./img/collFill1.png">

<h2>협업 필터링의 유형</h2>

* 최근접 이웃 기반(Nearest Neighbor)
-사용자 기반(User-user CF)
-아이템 기반(Item-item CF)

* 잠재 요인 기반(Latent Factor)
-행렬 분해 기반(Matrix Factorizatio)

<img src="./img/CollFill2.png">

<h2>협업 필터링의 특징</h2>

사용자 행동 데이터에 기반을 한다. 유저가 행으로 되어있고 아잍메이 열로 되어있는 데이터로 만들면 굉장히 쉽다.

평가자는 데이터들을 예측하는것이다. 평점이 좋으면 추천이고 평가를 해봤는데 평점이 낮으면 추천을 안하는거다.

<img src="./img/CollFill3.png">

<h2>협업 필터링을 위한 데이터 세트 - 사용자 로우- 아이템 컬럼</h2>

어떻게하는거냐면 먼저 데이터 세트를 변환해줘야된다.

일반적인 데이터 세트는 왼쪽같이 되어 있을 것임. 그런데 오른쪽으로 사용자가 행, 아이템이 열인 기반 데이터 셋.

일반적인 데이터는 사용자가 아이템 하나를 사고 레이팅 점수 하나. 이런식인데 바꿔주는거다.

엑셀 잘 사용하는 사람은 이거 피봇이네 라고 생각한다. 행 데이터를 칼럼으로 바꾸는 것임.

<img src="./img/UserRow.png">

<h2>사용자 기반과 아이템 기반 협업 필터링 이해</h2>

사용자 기반(User-User): 비슷한 고객을 기반으로 해서 비슷한 고객이 선택한. 

아이템 기반(Item-Item): 약간 헷갈릴 수 있다. 비슷한 아이템에 기반하는건 맞는데, 비슷하다는걸 어떻게할거냐? 앞에 같이 컨텐츠 기반으로 할거냐? 그게 아니라. 가령 아이템 I라는게 있다. 사용자 A,B,C에게 좋은 평점을 받았다. J라는 아이템이 있는데 또 좋은 평점을 받았다.
K라는 아이템이 있는데 이건 평점이 낮다. 그렇다면 I와 J를 비슷하다고 판단하는 것이다.

<img src="./img/UserItem1.png">

<h2>사용자 기반 협업 필터링</h2>

사용자 item 행렬 그대로 적용이 된다. A라는 사용자를 보면 다크나이트 인터스텔라 엣지오브 투모로우를 재밌게보고 프로메테우스, 스타워즈를 안봤다.

B하고 A가 유사한지를 본다. C라는 사용자를 봤더니 조금 냉소적이다. 그렇기 때문에 상호간에 유사도는 A와 B사용자가 유사하다고 보고, 

B 사용자가 높은 평점을 준 프로메테우스를 추천하는 것이다.

<img src="./img/User1.png">

<h2>아이템 기반 협업 필터링</h2>

굉장히 유사하다. 근데 앞에랑 데이터세트가 확 바뀌었다. 행이 사용자였는데 아이템으로 바뀌고 열이 사용자로 바뀌었다.

다크나이트를 보면 5,4,5,5,5

프로메테우스는 5,4,4, ,5

다크나이트를 재밌게 봤다면 프로메테우스를 추천하는 거다. 왜냐면 다크나이트 벡터는 프로메테우스 벡터에 더 유사하기 떄문이다.

<img src="./img/item1.png">

<h2>사용자 기반 VS 아이템 기반</h2>

일반적으로는 아이템 기반을 더 많이 쓴다. 같은 상품을 구매했다고 비슷한 상품이라고 볼 수 있느냐에 의구심이 들기 때문.

<img src="./img/User_vs_Item.png">

<h2>협업 필터링을 위한 코사인 유사도</h2>

협업 필터링 역시 유사도를 추출하는 방법은 코사인 유사도를 적용할 것임

<img src="./img/cosine1.png">

<h2>사용자 기반과 아이템 기반 간의 데이터 세트 변환</h2>

그리고 잠간 봤지만 사용자가 행 아이템이 열인 일반적인 메트릭스를 

아이템 기준으로 사용자가 열인 방식으로 바꿀건데 

간단하게 판다스의 transpose()를 이용할 것임

<img src="./img/dataset.png">

<h2>아이템 기반 협업 필터링의 개인화된 영화 추천</h2>

그리고 이렇게 유사도를 구했으면 유사도에 따라서 평점을 줘야한다. 앞에서도 컨텐츠 할 때 유사도만 구하지 않고 평점에 따라서 필터링을 했다.

마찬가지로 유사도가 주어지면 평점에 따라서 가중치를 부여할 수 있다. 어떻게 하냐면

각각의 아이템간 유사도와 평점 데이터를 곱해준다. 그리고 그 합을 구하는거다. 뒤에는 정규화라고 생각하면 된다.

5,4,1,3,2로 줬다면 아이템 i와 jklmn의 유사도를 보니 유사도가 높은 아이템(i,l)일수록 그 아이템에 주어진 평점을

잘 반영해야 할 것이다. 그래서 그것을 그냥 곱해버린다. 그러면 유사도가 높은 아이템에 대해서 평점의 영향도가 커질 것이다.

여기서는 평점을 낮게 줬으니까 전반적으로 평점의 영향을 크게 받을 것임. 만약에 l의 평점을5점으로 바꿨다면 4.1이 되어버릴 것임.

그렇게 구하는 공식이 평점 벡터 곱하기 유사도 벡터이다. 실습할때 구현을 해볼 것이다.

<img src="./img/movierecom1.png">

<h2>아이템 기반 협업 필터링의 개인화된 영화 추천</h2>

i와 유사한 n개의 아이템에 대해서 유사도를 구하고 

top_n개의 아이템에 대해서 개별 평점들을 다 곱해서 최종합을 만드는 거다. 그 다음에 정규화를 만드는 것(앞의 식의 정리)

<img src="./img/itemMovRecom.png">

<h2>아이템 기반 협업 필터링 구현 순서</h2>

1. 사용자-아이템 행렬 데이터를 아이템- 사용자 행렬 데이터로 변환

2. 아이템간의 코사인 유사도로 아이템 유사도 산출

3. 사용자가 관람(구매)하지 않은 아이템들 중에서 아이템간 유사도를 반영한 예측 점수 계산

4. 예측 점수가 가장 높은 순으로 아이템 추천

<img src="./img/itemImpleorder.png">

<h2>아이템 기반 협업 필터링 실습 -MovieLens 데이터 세트</h2>

굉장히 유명한 데이터 세트임. 그룹렌즈라는 사이트에서 만들었다.

파일을 받았을때 한번씩 열어보면서 확인 해보자.

groplens mivelens dataset에서 만듦. 다섯개의 파일중 movies랑 ratings를 주로 쓸거다.

ratings를 주로 쓸건데 userid와 movieid, rating을 주로 쓸것임.

이렇게 row레벨로 되어있는 데이터세트를 어떻게 해야한다고 했던가?

사용자가 행, 아이템이 열로 되어있는 걸로 바꿔야한다.

movies에서는 타이틀만 쓸것임.

두 파일을 조인할거다.

<img src="./img/MoviesLens1.png"/>

**데이터 로드**

In [1]:
import pandas as pd
import numpy as np

movies = pd.read_csv('./ml-latest-small/movies.csv')
ratings = pd.read_csv('./ml-latest-small/ratings.csv')
print(movies.shape)
print(ratings.shape)

(9742, 3)
(100836, 4)


위 결과를 보니 영화는 9742건, ratings는 100836건이 있다.<br>
데이터 샘플을 한번 보자

In [2]:
movies.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [3]:
ratings.head()

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

**로우레벨 사용자 평점 데이터를 사용자-아이템 평점 행렬로 변환**<br><br>
timestamp는 필요없으니까 필터링해서 데이터프레임으로 다시 만들고 변환하자.  
pivot_table을 사용할것임.

In [4]:
ratings = ratings[['userId', 'movieId', 'rating']]

# index, 즉 행은 userID로, columns, 즉 열은 movieID로 할 것임. value는 레이팅이다.(쉬프트+탭으로 확인해보기)
ratings_matrix = ratings.pivot_table('rating', index='userId', columns='movieId')
ratings_matrix.head(3)


movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           4.0     NaN     4.0     NaN     NaN     4.0     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  9       10      ...  193565  193567  193571  193573  193579  193581  \
userId                   ...                                                   
1           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  193583  193585  193587  193609  
userId                                   
1           NaN     NaN     NaN     NaN  
2           NaN     NaN     NaN     NaN  
3           NaN     NaN     NaN     NaN  

[3 rows x 9724 columns]

위 결과를 확인해보니 movie가 숫자로 되어있다.

사용자는 숫자로 되어도 그렇게 상관이 없는데 movieID는 숫자로 되어있으면 가독성이 좀 떨어진다.



아까 무비ID 별로 영화명을 가지고 있는 데이터 프레임이 movies였다.

그래서 데이터프레임을 조인할 것임.

In [11]:
# 데이터를 미리 한번 볼까요?
rating_movies = pd.merge(ratings, movies, on='movieId')

In [12]:
rating_movies.head()

userId  movieId  rating             title  \
0       1        1     4.0  Toy Story (1995)   
1       5        1     4.0  Toy Story (1995)   
2       7        1     4.5  Toy Story (1995)   
3      15        1     2.5  Toy Story (1995)   
4      17        1     4.5  Toy Story (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1  Adventure|Animation|Children|Comedy|Fantasy  
2  Adventure|Animation|Children|Comedy|Fantasy  
3  Adventure|Animation|Children|Comedy|Fantasy  
4  Adventure|Animation|Children|Comedy|Fantasy

위 결과를 보니 조인이 된 컬럼까지 같이 붙어있다.<br>

즉 rating_movies를 어떻게 바꿀거냐면 타이틀로 바꿀거다. 그리고 나머지는 똑같다.

그리고 NaN의 값도 0으로 바꿀거다. 이게 최소값이 0.5고 최대값이 5다. 그래서 NaN 값을 0으로 바꿀거고

문맥적으론 null값이다. 그러니까 아직 사용자가 평가하지 않은 값이라는 것임

In [13]:
# title 컬럼을 얻기 이해 movies 와 조인 수행
rating_movies = pd.merge(ratings, movies, on='movieId')

# columns='title' 로 title 컬럼으로 pivot 수행. 
ratings_matrix = rating_movies.pivot_table('rating', index='userId', columns='title')

# NaN 값을 모두 0 으로 변환
ratings_matrix = ratings_matrix.fillna(0)
ratings_matrix.head(3)


title   '71 (2014)  'Hellboy': The Seeds of Creation (2004)  \
userId                                                        
1              0.0                                      0.0   
2              0.0                                      0.0   
3              0.0                                      0.0   

title   'Round Midnight (1986)  'Salem's Lot (2004)  \
userId                                                
1                          0.0                  0.0   
2                          0.0                  0.0   
3                          0.0                  0.0   

title   'Til There Was You (1997)  'Tis the Season for Love (2015)  \
userId                                                               
1                             0.0                              0.0   
2                             0.0                              0.0   
3                             0.0                              0.0   

title   'burbs, The (1989)  'night Mother (1986)  (500) Days of Summer (2009)  \
userId                                                                          
1                      0.0                   0.0                          0.0   
2                      0.0                   0.0                          0.0   
3                      0.0                   0.0                          0.0   

title   *batteries not included (1987)  ...  Zulu (2013)  [REC] (2007)  \
userId                                  ...                              
1                                  0.0  ...          0.0           0.0   
2                                  0.0  ...          0.0           0.0   
3                                  0.0  ...          0.0           0.0   

title   [REC]² (2009)  [REC]³ 3 Génesis (2012)  \
userId                                           
1                 0.0                      0.0   
2                 0.0                      0.0   
3                 0.0                      0.0   

title   anohana: The Flower We Saw That Day - The Movie (2013)  \
userId                                                           
1                                                     0.0        
2                                                     0.0        
3                                                     0.0        

title   eXistenZ (1999)  xXx (2002)  xXx: State of the Union (2005)  \
userId                                                                
1                   0.0         0.0                             0.0   
2                   0.0         0.0                             0.0   
3                   0.0         0.0                             0.0   

title   ¡Three Amigos! (1986)  À nous la liberté (Freedom for Us) (1931)  
userId                                                                    
1                         4.0                                        0.0  
2                         0.0                                        0.0  
3                         0.0                                        0.0  

[3 rows x 9719 columns]

위의 결과를보자. 컬럼명으로 열이 되어있는걸 확인할 수 있다.

그리고 보면 0이 많아서 가독성이 많이 떨어진다고 생각할 수 있는건데 이게 맞다. 왜냐면

사용자 item 행렬은 희소행렬이다. 아무리 영화광이더라도 1만개의 영화를 다 개인이 볼 수 가 없다.

상품도 마찬가지다. 다 평점을 주거나 그럴 수는 없다. 그러므로 추천에 쓰이는 데이터는 기본적으로 희소행렬이다.

자 그런데, 사용자 아이템 행렬이 아니라 뭘로 바꿔야 아이템 기반 협업 필터링이 됐었나??

앞장을 보면 

 **아이템 기반 협업 필터링 구현 순서**

1. 사용자- 아이템 행렬 데이터를 아이템-사용자 행렬 데이터로 변환

2. 아이템간의 코사인 유사도로 아이템 유사도 산출

3. 사용자가 관람(구매)하지 않은 아이템들 중에서 아잍메간 유사도를 반영한 예측 점수 계산

4. 예측 점수가 가장 높은 순으로 아이템 추천

이렇게 공부했다. 아이템이 횡으로 바껴야 한다. 아이템으로 바꼈으니까 코사인 유사도를 쉽게 구할 수 있다.

아이템 횡으로 바꾸는걸 가장 쉽게할 수 있는게 뭐였냐면 transpose()다.

### 영화와 영화들 간 유사도 산출

In [14]:
# transpose()를 쓰면 열과 행을 바꿀 수 있다.
ratings_matrix_T = ratings_matrix.transpose()
ratings_matrix_T.head(3)

userId                                   1    2    3    4    5    6    7    \
title                                                                        
'71 (2014)                               0.0  0.0  0.0  0.0  0.0  0.0  0.0   
'Hellboy': The Seeds of Creation (2004)  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
'Round Midnight (1986)                   0.0  0.0  0.0  0.0  0.0  0.0  0.0   

userId                                   8    9    10   ...  601  602  603  \
title                                                   ...                  
'71 (2014)                               0.0  0.0  0.0  ...  0.0  0.0  0.0   
'Hellboy': The Seeds of Creation (2004)  0.0  0.0  0.0  ...  0.0  0.0  0.0   
'Round Midnight (1986)                   0.0  0.0  0.0  ...  0.0  0.0  0.0   

userId                                   604  605  606  607  608  609  610  
title                                                                       
'71 (2014)                               0.0  0.0  0.0  0.0  0.0  0.0  4.0  
'Hellboy': The Seeds of Creation (2004)  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
'Round Midnight (1986)                   0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[3 rows x 610 columns]

In [15]:
ratings_matrix_T.shape

(9719, 610)

앞에 9742개였는데 왜 9719개냐고 할 수 있는데 join이 되면 몇개가 빠진다. 데이터가 없는 것들이라던가.

이번에는 cosine_similarity를 이용해서 ratings_matrix_T의  아이템 사용자 평점 행렬의 코사인 유사도를 구할거다. 

그러면 (9719,9719)의 쉐이프가 나와야 될 것이다. 

item_sim을 dataFrame으로 만들거다. 

In [16]:
from sklearn.metrics.pairwise import cosine_similarity

item_sim = cosine_similarity(ratings_matrix_T, ratings_matrix_T)

# cosine_similarity() 로 반환된 넘파이 행렬을 영화명을 매핑하여 DataFrame으로 변환
item_sim_df = pd.DataFrame(data=item_sim, index=ratings_matrix.columns,
                          columns=ratings_matrix.columns)
print(item_sim_df.shape)
item_sim_df.head(3)


(9719, 9719)


title                                    '71 (2014)  \
title                                                 
'71 (2014)                                      1.0   
'Hellboy': The Seeds of Creation (2004)         0.0   
'Round Midnight (1986)                          0.0   

title                                    'Hellboy': The Seeds of Creation (2004)  \
title                                                                              
'71 (2014)                                                              0.000000   
'Hellboy': The Seeds of Creation (2004)                                 1.000000   
'Round Midnight (1986)                                                  0.707107   

title                                    'Round Midnight (1986)  \
title                                                             
'71 (2014)                                             0.000000   
'Hellboy': The Seeds of Creation (2004)                0.707107   
'Round Midnight (1986)                                 1.000000   

title                                    'Salem's Lot (2004)  \
title                                                          
'71 (2014)                                               0.0   
'Hellboy': The Seeds of Creation (2004)                  0.0   
'Round Midnight (1986)                                   0.0   

title                                    'Til There Was You (1997)  \
title                                                                
'71 (2014)                                                     0.0   
'Hellboy': The Seeds of Creation (2004)                        0.0   
'Round Midnight (1986)                                         0.0   

title                                    'Tis the Season for Love (2015)  \
title                                                                      
'71 (2014)                                                           0.0   
'Hellboy': The Seeds of Creation (2004)                              0.0   
'Round Midnight (1986)                                               0.0   

title                                    'burbs, The (1989)  \
title                                                         
'71 (2014)                                         0.000000   
'Hellboy': The Seeds of Creation (2004)            0.000000   
'Round Midnight (1986)                             0.176777   

title                                    'night Mother (1986)  \
title                                                           
'71 (2014)                                                0.0   
'Hellboy': The Seeds of Creation (2004)                   0.0   
'Round Midnight (1986)                                    0.0   

title                                    (500) Days of Summer (2009)  \
title                                                                  
'71 (2014)                                                  0.141653   
'Hellboy': The Seeds of Creation (2004)                     0.000000   
'Round Midnight (1986)                                      0.000000   

title                                    *batteries not included (1987)  ...  \
title                                                                    ...   
'71 (2014)                                                          0.0  ...   
'Hellboy': The Seeds of Creation (2004)                             0.0  ...   
'Round Midnight (1986)                                              0.0  ...   

title                                    Zulu (2013)  [REC] (2007)  \
title                                                                
'71 (2014)                                       0.0      0.342055   
'Hellboy': The Seeds of Creation (2004)          0.0      0.000000   
'Round Midnight (1986)                           0.0      0.000000   

title                                    [REC]² (2009)  \
title                                                    
'71 (2014)                                 

In [17]:
# Godfather하고 유사한거 6개만 해보자
item_sim_df["Godfather, The (1972)"].sort_values(ascending=False)[:6]

title
Godfather, The (1972)                        1.000000
Godfather: Part II, The (1974)               0.821773
Goodfellas (1990)                            0.664841
One Flew Over the Cuckoo's Nest (1975)       0.620536
Star Wars: Episode IV - A New Hope (1977)    0.595317
Fargo (1996)                                 0.588614
Name: Godfather, The (1972), dtype: float64

In [18]:
# 인셉션도 해보자. 자기꺼 빼고 해보자
item_sim_df["Inception (2010)"].sort_values(ascending=False)[1:6]

title
Dark Knight, The (2008)          0.727263
Inglourious Basterds (2009)      0.646103
Shutter Island (2010)            0.617736
Dark Knight Rises, The (2012)    0.617504
Fight Club (1999)                0.615417
Name: Inception (2010), dtype: float64

이제 유사도에 기반해서 유사도 값에 가중치를 줘서 예측 평점을 할거다. 

그 식은 

<img src="./img/itemMovRecom.png"/>

### 아이템 기반 인접 이웃 협업 필터링으로 개인화된 영화 추천

In [19]:
# 유사도에다 가중 평균을 두면서 예측하는 함수를 만들 것임.
# ratings_arr 은 앞에서 구한 ratings_matrix이다. item_sim_arr 은 아이템간 코사인 유사도이다.
def predict_rating(ratings_arr, item_sim_arr ):
    ratings_pred = ratings_arr.dot(item_sim_arr)/ np.array([np.abs(item_sim_arr).sum(axis=1)]) #뒤에 나눈 부분은 정규화를 위해 유사도를 다 더한 것임
    return ratings_pred


In [21]:
item_sim_df.head(3)

title                                    '71 (2014)  \
title                                                 
'71 (2014)                                      1.0   
'Hellboy': The Seeds of Creation (2004)         0.0   
'Round Midnight (1986)                          0.0   

title                                    'Hellboy': The Seeds of Creation (2004)  \
title                                                                              
'71 (2014)                                                              0.000000   
'Hellboy': The Seeds of Creation (2004)                                 1.000000   
'Round Midnight (1986)                                                  0.707107   

title                                    'Round Midnight (1986)  \
title                                                             
'71 (2014)                                             0.000000   
'Hellboy': The Seeds of Creation (2004)                0.707107   
'Round Midnight (1986)                                 1.000000   

title                                    'Salem's Lot (2004)  \
title                                                          
'71 (2014)                                               0.0   
'Hellboy': The Seeds of Creation (2004)                  0.0   
'Round Midnight (1986)                                   0.0   

title                                    'Til There Was You (1997)  \
title                                                                
'71 (2014)                                                     0.0   
'Hellboy': The Seeds of Creation (2004)                        0.0   
'Round Midnight (1986)                                         0.0   

title                                    'Tis the Season for Love (2015)  \
title                                                                      
'71 (2014)                                                           0.0   
'Hellboy': The Seeds of Creation (2004)                              0.0   
'Round Midnight (1986)                                               0.0   

title                                    'burbs, The (1989)  \
title                                                         
'71 (2014)                                         0.000000   
'Hellboy': The Seeds of Creation (2004)            0.000000   
'Round Midnight (1986)                             0.176777   

title                                    'night Mother (1986)  \
title                                                           
'71 (2014)                                                0.0   
'Hellboy': The Seeds of Creation (2004)                   0.0   
'Round Midnight (1986)                                    0.0   

title                                    (500) Days of Summer (2009)  \
title                                                                  
'71 (2014)                                                  0.141653   
'Hellboy': The Seeds of Creation (2004)                     0.000000   
'Round Midnight (1986)                                      0.000000   

title                                    *batteries not included (1987)  ...  \
title                                                                    ...   
'71 (2014)                                                          0.0  ...   
'Hellboy': The Seeds of Creation (2004)                             0.0  ...   
'Round Midnight (1986)                                              0.0  ...   

title                                    Zulu (2013)  [REC] (2007)  \
title                                                                
'71 (2014)                                       0.0      0.342055   
'Hellboy': The Seeds of Creation (2004)          0.0      0.000000   
'Round Midnight (1986)                           0.0      0.000000   

title                                    [REC]² (2009)  \
title                                                    
'71 (2014)                                 

In [20]:
ratings_matrix.head(3)

title   '71 (2014)  'Hellboy': The Seeds of Creation (2004)  \
userId                                                        
1              0.0                                      0.0   
2              0.0                                      0.0   
3              0.0                                      0.0   

title   'Round Midnight (1986)  'Salem's Lot (2004)  \
userId                                                
1                          0.0                  0.0   
2                          0.0                  0.0   
3                          0.0                  0.0   

title   'Til There Was You (1997)  'Tis the Season for Love (2015)  \
userId                                                               
1                             0.0                              0.0   
2                             0.0                              0.0   
3                             0.0                              0.0   

title   'burbs, The (1989)  'night Mother (1986)  (500) Days of Summer (2009)  \
userId                                                                          
1                      0.0                   0.0                          0.0   
2                      0.0                   0.0                          0.0   
3                      0.0                   0.0                          0.0   

title   *batteries not included (1987)  ...  Zulu (2013)  [REC] (2007)  \
userId                                  ...                              
1                                  0.0  ...          0.0           0.0   
2                                  0.0  ...          0.0           0.0   
3                                  0.0  ...          0.0           0.0   

title   [REC]² (2009)  [REC]³ 3 Génesis (2012)  \
userId                                           
1                 0.0                      0.0   
2                 0.0                      0.0   
3                 0.0                      0.0   

title   anohana: The Flower We Saw That Day - The Movie (2013)  \
userId                                                           
1                                                     0.0        
2                                                     0.0        
3                                                     0.0        

title   eXistenZ (1999)  xXx (2002)  xXx: State of the Union (2005)  \
userId                                                                
1                   0.0         0.0                             0.0   
2                   0.0         0.0                             0.0   
3                   0.0         0.0                             0.0   

title   ¡Three Amigos! (1986)  À nous la liberté (Freedom for Us) (1931)  
userId                                                                    
1                         4.0                                        0.0  
2                         0.0                                        0.0  
3                         0.0                                        0.0  

[3 rows x 9719 columns]

In [22]:
ratings_pred = predict_rating(ratings_matrix.values , item_sim_df.values)

# 위에걸 다시 데이터 프레임으로 만들었음
ratings_pred_matrix = pd.DataFrame(data=ratings_pred, index= ratings_matrix.index,
                                   columns = ratings_matrix.columns)
print(ratings_pred_matrix.shape)
ratings_pred_matrix.head(3)

(610, 9719)


title   '71 (2014)  'Hellboy': The Seeds of Creation (2004)  \
userId                                                        
1         0.070345                                 0.577855   
2         0.018260                                 0.042744   
3         0.011884                                 0.030279   

title   'Round Midnight (1986)  'Salem's Lot (2004)  \
userId                                                
1                     0.321696             0.227055   
2                     0.018861             0.000000   
3                     0.064437             0.003762   

title   'Til There Was You (1997)  'Tis the Season for Love (2015)  \
userId                                                               
1                        0.206958                         0.194615   
2                        0.000000                         0.035995   
3                        0.003749                         0.002722   

title   'burbs, The (1989)  'night Mother (1986)  (500) Days of Summer (2009)  \
userId                                                                          
1                 0.249883              0.102542                     0.157084   
2                 0.013413              0.002314                     0.032213   
3                 0.014625              0.002085                     0.005666   

title   *batteries not included (1987)  ...  Zulu (2013)  [REC] (2007)  \
userId                                  ...                              
1                             0.178197  ...     0.113608      0.181738   
2                             0.014863  ...     0.015640      0.020855   
3                             0.006272  ...     0.006923      0.011665   

title   [REC]² (2009)  [REC]³ 3 Génesis (2012)  \
userId                                           
1            0.133962                 0.128574   
2            0.020119                 0.015745   
3            0.011800                 0.012225   

title   anohana: The Flower We Saw That Day - The Movie (2013)  \
userId                                                           
1                                                0.006179        
2                                                0.049983        
3                                                0.000000        

title   eXistenZ (1999)  xXx (2002)  xXx: State of the Union (2005)  \
userId                                                                
1              0.212070    0.192921                        0.136024   
2              0.014876    0.021616                        0.024528   
3              0.008194    0.007017                        0.009229   

title   ¡Three Amigos! (1986)  À nous la liberté (Freedom for Us) (1931)  
userId                                                                    
1                    0.292955                                   0.720347  
2                    0.017563                                   0.000000  
3                    0.010420                                   0.084501  

[3 rows x 9719 columns]

**가중치 평점 부여뒤에 예측 성능 평가 MSE를 구함**

평가를 위해서.

In [23]:
from sklearn.metrics import mean_squared_error

# 사용자가 평점을 부여한 영화에 대해서만 예측 성능 평가 MSE 를 구함. 
def get_mse(pred, actual):
    # Ignore nonzero terms. 0은 NaN인데 이것도 다 평가를 한 것이므로 뺄 필요가 있다.
    pred = pred[actual.nonzero()].flatten()
    actual = actual[actual.nonzero()].flatten()
    return mean_squared_error(pred, actual)

print('아이템 기반 모든 인접 이웃 MSE: ', get_mse(ratings_pred, ratings_matrix.values ))


아이템 기반 모든 인접 이웃 MSE:  9.895354759094706


예측 평점이 실제평점하고는 다르다. 가중치를 포함해서 계산한 거기 때문에 당연히 오류가 클 수 밖에 없다.

이것들을 이제 top-n 유사도를 가진 데이터들에 대해서만 예측 평점을 계산해본다.

**top-n 유사도를 가진 데이터들에 대해서만 예측 평점 계산**

이게 살짝 헷갈릴 수 있다.

In [24]:
def predict_rating_topsim(ratings_arr, item_sim_arr, n=20):
    # 사용자-아이템 평점 행렬 크기만큼 0으로 채운 예측 행렬 초기화
    pred = np.zeros(ratings_arr.shape)

    # 사용자-아이템 평점 행렬의 열 크기만큼 Loop 수행.  컬럼 별로 먼저 돌거다.
    for col in range(ratings_arr.shape[1]):
        # 유사도 행렬에서 유사도가 큰 순으로 n개 데이터 행렬의 index 반환. 
        top_n_items = [np.argsort(item_sim_arr[:, col])[:-n-1:-1]]
        # 개인화된 예측 평점을 계산. 이제 행으로 다시 루프를 돌린다.
        for row in range(ratings_arr.shape[0]):
            pred[row, col] = item_sim_arr[col, :][top_n_items].dot(ratings_arr[row, :][top_n_items].T) 
            pred[row, col] /= np.sum(np.abs(item_sim_arr[col, :][top_n_items]))        
    return pred


**top-n 유사도 기반의 예측 평점 및 MSE 계산**

In [25]:
ratings_pred = predict_rating_topsim(ratings_matrix.values , item_sim_df.values, n=20)
print('아이템 기반 인접 TOP-20 이웃 MSE: ', get_mse(ratings_pred, ratings_matrix.values ))


# 계산된 예측 평점 데이터는 DataFrame으로 재생성
ratings_pred_matrix = pd.DataFrame(data=ratings_pred, index= ratings_matrix.index,
                                   columns = ratings_matrix.columns)


C:\Users\esaw2\Anaconda3\envs\pygta5\lib\site-packages\ipykernel_launcher.py:11: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  # This is added back by InteractiveShellApp.init_path()
C:\Users\esaw2\Anaconda3\envs\pygta5\lib\site-packages\ipykernel_launcher.py:12: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  if sys.path[0] == '':


아이템 기반 인접 TOP-20 이웃 MSE:  3.695009387428144


MSE는 3.69로 줄었다. 당연히 줄어드는 것임.

이제는 ratings_pred_matrix에서 나온게 최종적인 예측평점 데이터이다. 아이템 기반 협업 필터링으로 만들어진.

이제는 여기 데이터에서 평점 데이터가 큰 것을 추출할 거다.

In [26]:
#특정 사용자 유저인덱스 9번의 데이터를 한번 확인해볼 것임.
user_rating_id = ratings_matrix.loc[9, :]
user_rating_id[ user_rating_id > 0].sort_values(ascending=False)[:10]

title
Adaptation (2002)                                                                 5.0
Austin Powers in Goldmember (2002)                                                5.0
Lord of the Rings: The Fellowship of the Ring, The (2001)                         5.0
Lord of the Rings: The Two Towers, The (2002)                                     5.0
Producers, The (1968)                                                             5.0
Citizen Kane (1941)                                                               5.0
Raiders of the Lost Ark (Indiana Jones and the Raiders of the Lost Ark) (1981)    5.0
Back to the Future (1985)                                                         5.0
Glengarry Glen Ross (1992)                                                        4.0
Sunset Blvd. (a.k.a. Sunset Boulevard) (1950)                                     4.0
Name: 9, dtype: float64

**사용자가 관람하지 않은 영화 중에서 아이템 기반의 인접 이웃 협업 필터링으로 영화 추천**<br>

위에서 본 9번 사용자에 대해서 추천할 것임.

관람하지 않은 영화를 추출할거다. 로직은 간단함.

In [27]:
def get_unseen_movies(ratings_matrix, userId):
    # userId로 입력받은 사용자의 모든 영화정보 추출하여 Series로 반환함. 
    # 반환된 user_rating 은 영화명(title)을 index로 가지는 Series 객체임. 
    user_rating = ratings_matrix.loc[userId,:]
    
    # user_rating이 0보다 크면 기존에 관람한 영화임. 대상 index를 추출하여 list 객체로 만듬
    already_seen = user_rating[ user_rating > 0].index.tolist()
    
    # 모든 영화명을 list 객체로 만듬. 
    movies_list = ratings_matrix.columns.tolist()
    
    # list comprehension으로 already_seen에 해당하는 movie는 movies_list에서 제외함. 
    unseen_list = [ movie for movie in movies_list if movie not in already_seen]
    
    return unseen_list


**아이템 기반 유사도로 평점이 부여된 데이터 세트에서 해당 사용자가 관람하지 않은 영화들의 예측 평점이 가장 높은 영화를 추천**

In [28]:
def recomm_movie_by_userid(pred_df, userId, unseen_list, top_n=10):
    # 예측 평점 DataFrame에서 사용자id index와 unseen_list로 들어온 영화명 컬럼을 추출하여
    # 가장 예측 평점이 높은 순으로 정렬함. 
    recomm_movies = pred_df.loc[userId, unseen_list].sort_values(ascending=False)[:top_n]
    return recomm_movies
    
# 사용자가 관람하지 않는 영화명 추출   
unseen_list = get_unseen_movies(ratings_matrix, 9)

# 아이템 기반의 인접 이웃 협업 필터링으로 영화 추천 
recomm_movies = recomm_movie_by_userid(ratings_pred_matrix, 9, unseen_list, top_n=10)

# 평점 데이타를 DataFrame으로 생성. 
recomm_movies = pd.DataFrame(data=recomm_movies.values,index=recomm_movies.index,columns=['pred_score'])
recomm_movies


pred_score
title                                                         
Shrek (2001)                                          0.866202
Spider-Man (2002)                                     0.857854
Last Samurai, The (2003)                              0.817473
Indiana Jones and the Temple of Doom (1984)           0.816626
Matrix Reloaded, The (2003)                           0.800990
Harry Potter and the Sorcerer's Stone (a.k.a. H...    0.765159
Gladiator (2000)                                      0.740956
Matrix, The (1999)                                    0.732693
Pirates of the Caribbean: The Curse of the Blac...    0.689591
Lord of the Rings: The Return of the King, The ...    0.676711

SF적이면서 장대한. 취향에 맞는 영화가 추천되었다.